### ISSUE TITLE
remove var_alloc_stack and associated pointers

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/math#565]
#### Summary:

* [ ] in `stan/math/rev/core/autodiffstackstorage.hpp`remove the member variable `AutodiffStackStorage::var_alloc_stack_`
* [ ] replace all the functions that depend on `var_alloc_stack_` with explicit allocations wrapped in `Map`
* [ ] add doc to the memory classes for reverse-mode autodiff

Also add documentation.

#### Description:

This will make the autodiff infrastructure simpler and a bit faster, because there'll only be two stacks to clean.

#### Current Version:
v2.16.0

**Comment by rok-cesnovar** [comment]
Is this issue still relevant? I am guessing with the recent refactors it is not?

**Comment by bbbales2** [comment]
I don't think this is gonna be feasible much longer or it's worth doing anymore. This is a useful feature. I think there's at least a thing in opencl that depends on this, the scoped chainable stack will depend on this, and we have this code computing a QR twice that could be avoided easily with another one of these https://github.com/stan-dev/math/blob/develop/stan/math/rev/fun/mdivide_left.hpp#L31 .
